scripts/github_pr/github_pr.rb

```diff
@@ -0,0 +1,405 @@
+#!/usr/bin/ruby
+
+require 'octokit'
+require 'optparse'
+require 'yaml'
+require 'json'
+
+class GithubAPI
+  def get_client
+    client = Octokit::Client.new(:netrc => true)
+    client.auto_paginate = true
+    client.login
+    client
+  end
+
+  def client
+    @@client ||= get_client
+  end
+
+  def current_rate_limit
+    client.rate_limit.remaining
+  end
+end
+
+
+class GithubClient
+  RESULT_MESSAGES = {
+    :success => 'succeeded',
+    :failure => 'failed',
+    :error   => 'has an error',
+    :pending => 'is pending',
+  }
+
+  def initialize(conf = {})
+    @config = conf
+  end
+
+  def organization
+    @config[:organization]
+  end
+
+  def repository
+    @config[:repository]
+  end
+
+  def org_repo
+    "#{organization}/#{repository}"
+  end
+
+  def context
+    @config[:context]
+  end
+
+  def set_status(commit, details)
+    result = details[:status].to_sym
+    GithubAPI.new.client.create_status(
+      org_repo,
+      commit,
+      result,
+      { :context => context,
+        :description => status_description(details[:message], result),
+        :target_url => details[:target_url].to_s
+      }
+    )
+  end
+
+  def status_description(description, status)
+    description ||= RESULT_MESSAGES[status]
+    description.to_s
+  end
+
+  def get_full_sha_status(sha)
+    status = GithubAPI.new.client.status(org_repo, sha)
+    status.statuses.select{ |s| s.context == context }.first rescue {}
+  end
+
+  def get_sha_status(sha)
+    get_full_sha_status(sha).state rescue ''
+  end
+
+  def get_pr(pr)
+    GithubAPI.new.client.pull_request(org_repo, pr)
+  end
+
+  def get_pr_info(pr)
+    get_pr(pr).to_attrs
+  end
+
+  def get_pr_latest_sha(pr)
+    get_pr(pr).head.sha rescue ''
+  end
+
+  def is_latest_sha?(pr, sha)
+    get_pr_latest_sha(pr) == sha
+  end
+
+  # state = ["open"|"closed"]
+  def get_all_pull_requests(state)
+     GithubAPI.new.client.pull_requests(org_repo, :state => state)
+  end
+end
+
+class GithubPRWorker
+  def initialize(base_parameters = {}, config = {})
+    @base_parameters = base_parameters
+    @config = config
+  end
+
+  def ghc_ctx(org, repo, ctx)
+    {
+      :org_repo => org + "/" + repo,
+      :organization => org,
+      :repository => repo,
+      :context => ctx,
+      :config_base_path => File.dirname(@base_parameters[:config]),
+    }
+  end
+
+  def process_list
+    @config["pr_processing"] || []
+  end
+
+  def status_filter_config(config)
+    # the config for the Status filter might be overridden via command line parameters
+    if (@base_parameters.has_key?(:mode) && @base_parameters[:mode].to_s.size > 0) then
+      config["status"] = @base_parameters[:mode]
+    end
+    config
+  end
+
+  def debug?
+    @base_parameters.has_key?(:debugfilterchain) && \
+      @base_parameters[:debugfilterchain] == true
+  end
+
+  def debug_filterchain(pulls, filter)
+    return unless debug?
+    return unless filter
+    puts "=" * 25
```

Comment:
I would create and use a Logger instance for such outputs.

Reply:
If it supported general logging, I would agree. But this is the only place that we print something like debug output and it is to help the user debug his own config. Does that qualify as debug? I would rather see it as a verbose mode. ... I know I called it debug in the first place, but thats from the user perspective, he wants to debug his filterchain config.